### PR TITLE
drivers: audio: add Audio Codec shell commands

### PR DESCRIPTION
**Pull request by blemouzy**
Add shell commands to start, stop and set properties of an Audio Codec device.

### CODE REVIEW

File: drivers/audio/codec_shell.c
Comment (plbossart) [inline drivers/audio/codec_shell.c:16]: it'd be interesting to clarify what 'start/stop' mean in the context of an audio codec
a) start the interface, e.g. I2S
b) power-up the analog parts
c) all of the above
d) all of the above + more?

Also why is this limited to playback?

File: drivers/audio/codec_shell.c
Comment (plbossart) [inline drivers/audio/codec_shell.c:47]: can you set a property on multiple channels at the same time? That's handy for quick global mute or volume changes on all channels.

Comment (blemouzy) [reply drivers/audio/codec_shell.c:47]: Yes, for example with the following command:

`codec set_prop <device> mute all 1`

File: drivers/audio/codec_shell.c
Comment (blemouzy) [reply drivers/audio/codec_shell.c:16]: > it'd be interesting to clarify what 'start/stop' mean in the context of an audio codec a) start the interface, e.g. I2S b) power-up the analog parts c) all of the above d) all of the above + more?

I use the same description as the one in [Audio Codec documentation](https://docs.zephyrproject.org/latest/hardware/peripherals/audio/codec.html#c.audio_codec_start_output).

Stopping audio codec is setting codec output to mute or high impedance.
Starting audio codec is setting codec output to unmute or play.

That not affects I2S part (which is the input of the audio codec).

Maybe I will propose an `audio` command later to manage the I2S part (play/stop audio data through I2S.

File: drivers/audio/codec_shell.c
Comment (plbossart) [reply drivers/audio/codec_shell.c:47]: ok

File: drivers/audio/codec_shell.c
Comment (plbossart) [reply drivers/audio/codec_shell.c:16]: thanks for the pointer, I wasn't aware there was an audio codec API. My comments are really at the API level, the same opens on what 'start' or 'stop' means, muting or high-impedance are radically different.

But if this PR is just adding a shell to an existing API I have no objection.

**Review by plbossart — DISMISSED** [review]

**Comment by kartben** [comment]
@lyakh @lgirdwood @marc-hb @kv2019i could you please help with reviewing this one?

**Comment by carlescufi** [comment]
> The code looks very readable but I can't help here sorry.
> 
> Adding cc: @fredoh9 , @aiChaoSONG, @bardliao others?

@kv2019i @lyakh @lgirdwood @fredoh9 @aiChaoSONG  could you please help review this?

**Comment by marc-hb** [comment]
@blemouzy while this particular PR is vendor-independent, I noticed some of your other contributions to Zephyr are NXP-specific.

PRs rarely exist in a complete vacuum so is there anyone you're collaborating with who can help review this PR?

Finding reviewers is not an exact science and the author must help.

PS: there is a git conflict to solve right now (I wish Github notified the author when this happens)

File: drivers/audio/CMakeLists.txt
Comment (dbaluta) [inline drivers/audio/CMakeLists.txt:7]: Can you please clarify what is an 'audio codec shell command'? Does Zephyr has a 'shell' like normal Linux has? Also if this is so, why do you place your code inside `drivers/audio`  where for what i have seen there are only audio codec implementations.

Comment (blemouzy) [reply drivers/audio/CMakeLists.txt:7]: > Does Zephyr has a 'shell' like normal Linux has?

Yes, Zephyr has a shell (`CONFIG_SHELL` + [official doc](https://docs.zephyrproject.org/latest/services/shell/index.html)).
It is super conveniant for test and debug.
For example with `audio` command, my hardware team colleague was able to measure board power consumption at different volumes easily.

> Also if this is so, why do you place your code inside `drivers/audio` where for what i have seen there are only audio codec implementations.

All other driver related shell commands are in driver folder:

```
drivers/adc/adc_shell.c
drivers/bbram/bbram_shell.c
drivers/can/can_shell.c
drivers/counter/counter_timer_shell.c
drivers/dac/dac_shell.c
drivers/edac/shell.c
drivers/eeprom/eeprom_shell.c
drivers/flash/flash_shell.c
drivers/fpga/fpga_shell.c
drivers/gpio/gpio_shell.c
drivers/hwinfo/hwinfo_shell.c
drivers/i2c/i2c_shell.c
drivers/led/led_shell.c
drivers/lora/shell.c
drivers/mdio/mdio_shell.c
drivers/modem/modem_shell.c
drivers/pcie/host/shell.c
drivers/pm_cpu_ops/psci_shell.c
drivers/pwm/pwm_shell.c
drivers/regulator/regulator_shell.c
drivers/rtc/rtc_shell.c
drivers/sensor/sensor_shell.c
drivers/sensor/shell_battery.c
drivers/smbus/smbus_shell.c
drivers/virtualization/virt_ivshmem_shell.c
drivers/w1/w1_shell.c
drivers/watchdog/wdt_shell.c
drivers/wifi/eswifi/eswifi_shell.c
```

File: drivers/audio/codec_shell.c
Comment (dbaluta) [inline drivers/audio/codec_shell.c:66]: no need for new line here.

File: drivers/audio/codec_shell.c
Comment (dbaluta) [inline drivers/audio/codec_shell.c:177]: you can go with `entry->syntax = (dev) ? : dev->name : NULL;

File: drivers/audio/codec_shell.c
Comment (dbaluta) [inline drivers/audio/codec_shell.c:57]: I think we should avoid magic numbers. Something like:

```
#define SHELL_ARG_DEVICE_INDEX 1
#define SHELL_ARG_PROPERTY_INDEX 2
```

**Comment by dbaluta** [comment]
@blemouzy mostly looking good! I have some minors comments. Will approve once the comments are addressed.

File: drivers/audio/codec_shell.c
Comment (blemouzy) [reply drivers/audio/codec_shell.c:177]: The same code is used in several (10) others shell files:

https://github.com/zephyrproject-rtos/zephyr/blob/6d67a56d5b5ddb3f744eb844e6b4f7d033be7a94/drivers/sensor/sensor_shell.c#L543

https://github.com/zephyrproject-rtos/zephyr/blob/6d67a56d5b5ddb3f744eb844e6b4f7d033be7a94/drivers/i2c/i2c_shell.c#L314

...

I propose you to leave it as it

File: drivers/audio/codec_shell.c
Comment (blemouzy) [reply drivers/audio/codec_shell.c:57]: The same trick is used in several (6) shell files:

https://github.com/zephyrproject-rtos/zephyr/blob/6d67a56d5b5ddb3f744eb844e6b4f7d033be7a94/drivers/eeprom/eeprom_shell.c#L17-L31

https://github.com/zephyrproject-rtos/zephyr/blob/6d67a56d5b5ddb3f744eb844e6b4f7d033be7a94/drivers/gpio/gpio_shell.c#L24-L36

...

I propose you to leave it as it

File: drivers/audio/codec_shell.c
Comment (blemouzy) [reply drivers/audio/codec_shell.c:65]: Done

**Review by dbaluta — APPROVED** [review]

**Comment by kartben** [comment]
> > The code looks very readable but I can't help here sorry.
> > Adding cc: @fredoh9 , @aiChaoSONG, @bardliao others?
> 
> @kv2019i @lyakh @lgirdwood @fredoh9 @aiChaoSONG could you please help review this?

Still needs another pair of eyes to make it in time for 3.5 -- anyone? thanks!

**Comment by marc-hb** [comment]
I don't know it well but there is a process:
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-review-escalation
https://docs.zephyrproject.org/latest/project/project_roles.html

**Review by nashif — APPROVED** [review]

**Pull request merged**
